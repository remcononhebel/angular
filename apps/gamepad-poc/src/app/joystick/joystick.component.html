<div class="joystick-container">
  <div class="joystick" [attr.id]="joystickId">&nbsp;</div>
</div>
