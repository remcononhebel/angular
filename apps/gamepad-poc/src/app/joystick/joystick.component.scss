.joystick-container {
  position: relative;
  border: 1px solid black;
  width: 200px;
  height: 200px;
  z-index: 1000;
}

.joystick {
  height: 100%;
}
