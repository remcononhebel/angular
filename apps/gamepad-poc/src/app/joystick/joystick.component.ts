import { AfterViewInit, Component, EventEmitter, Input, Output } from '@angular/core';
import nipplejs from 'nipplejs';

@Component({
  selector: 'app-joystick',
  standalone: true,
  imports: [],
  templateUrl: './joystick.component.html',
  styleUrl: './joystick.component.scss',
})
export class JoystickComponent implements AfterViewInit {
  @Input() joystickId!: string;
  @Input() options!: nipplejs.JoystickManagerOptions;

  @Output() move = new EventEmitter<nipplejs.JoystickOutputData>();

  private manager!: nipplejs.JoystickManager;

  ngAfterViewInit(): void {
    const options: nipplejs.JoystickManagerOptions = { ...this.options, zone: this.getJoystickElement(this.joystickId) };

    this.manager = nipplejs.create(options);
    this.manager.on('move', (_: unknown, data: nipplejs.JoystickOutputData) => {
      this.move.emit(data);
    });
  }

  getJoystickElement(joystickId: string): HTMLElement | undefined {
    return document.getElementById(joystickId) ?? undefined;
  }
}
