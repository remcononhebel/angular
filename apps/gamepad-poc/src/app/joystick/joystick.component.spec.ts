import { ComponentFixture } from '@angular/core/testing';
import { SpectatorHost, createHostFactory } from '@ngneat/spectator';
import nipplejs, { JoystickOutputData } from 'nipplejs';
import { JoystickComponent } from './joystick.component';

jest.mock('nipplejs', () => {
  return {
    create: jest.fn().mockReturnValue({
      on: (_: string, callback: (evt: nipplejs.EventData, data: nipplejs.JoystickOutputData) => void) =>
        callback({} as nipplejs.EventData, {} as JoystickOutputData),
    }),
  };
});

describe('JoystickComponent', () => {
  let spectator: SpectatorHost<JoystickComponent>;
  let component: JoystickComponent;
  let fixture: ComponentFixture<JoystickComponent>;
  const createHost = createHostFactory({ component: JoystickComponent, imports: [] });

  const mockJoystickId = 'mockJoystickId';

  beforeEach(() => {
    spectator = createHost(`<app-joystick [joystickId]="'${mockJoystickId}'" [options]="{}"></app-joystick>`);
    component = spectator.component;
    fixture = spectator.fixture;
  });

  it('should create component', () => {
    expect(component).toBeTruthy();
  });

  it('should render a joystick', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('getJoystickElement', () => {
    it('should return undefined if element cannot be found', () => {
      expect(component.getJoystickElement('does not exist')).toBeFalsy();
    });

    it('should return undefined if element cannot be found', () => {
      expect(component.getJoystickElement(mockJoystickId)).toBeTruthy();
    });
  });
});
