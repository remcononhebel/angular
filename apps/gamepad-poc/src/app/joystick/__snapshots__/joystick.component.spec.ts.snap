// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoystickComponent should render a joystick 1`] = `
<ng-component>
  <app-joystick>
    <div
      class="joystick-container"
    >
      <div
        class="joystick"
        id="mockJoystickId"
      >
      </div>
    </div>
  </app-joystick>
</ng-component>
`;
